distributed/tests/test_cancelled_state.py

```diff
@@ -240,3 +235,133 @@ async def get_data(self, comm, *args, **kwargs):
+        fut1 = c.submit(
+            block_execution,
+            1,
+            executing,
+            lock_executing,
+            workers=[a.address],
+            allow_other_workers=True,
+        )
+        await fut1
+        await lock_executing.acquire()
+        in_get_data.clear()
+        await block_get_data.acquire()
+        fut2 = c.submit(inc, fut1, workers=[b.address])
+
+        await in_get_data.wait()
+        await s.remove_worker(a.address, "foo", close=False, safe=True)
+
+        while not b.tasks[fut1.key].state == "resumed":
+            await asyncio.sleep(0.1)
+
+        fut1.release()
+        fut2.release()
+        await asyncio.sleep(0.2)
+        block_get_data.release()
+        while b.tasks:
+            await asyncio.sleep(0.1)
+
+    assert_story(
+        b.story(fut1.key),
+        expect=[
+            # The initial free-keys is rejected
+            ("free-keys", (fut1.key,)),
+            (fut1.key, "resumed", "released", "cancelled", {}),
+            # After gather_dep receives the data, it tries to transition to memory but the task will release instead
+            (fut1.key, "cancelled", "memory", "released", {fut1.key: "forgotten"}),
+        ],
+    )
+
+
+@gen_cluster(client=True)
+async def test_cancelled_error(c, s, a, b):
+    executing = Event()
+    lock_executing = Lock()
+    await lock_executing.acquire()
+
+    def block_execution(x, event, lock):
+        event.set()
+        with lock:
+            raise RuntimeError()
+
+    fut1 = c.submit(
+        block_execution,
+        1,
+        executing,
+        lock_executing,
+        workers=[b.address],
+        allow_other_workers=True,
+    )
+
+    await executing.wait()
+    assert b.tasks[fut1.key].state == "executing"
+    fut1.release()
+    while b.tasks[fut1.key].state == "executing":
+        await asyncio.sleep(0.1)
+    await lock_executing.release()
+    while b.tasks:
+        await asyncio.sleep(0.1)
+
+    assert_story(
+        b.story(fut1.key),
+        [
+            (fut1.key, "executing", "released", "cancelled", {}),
+            (fut1.key, "cancelled", "error", "error", {fut1.key: "released"}),
+        ],
+    )
+
+
+@gen_cluster(client=True, nthreads=[("", 1, {"resources": {"A": 1}})])
+async def test_cancelled_error_with_ressources(c, s, a):
+    executing = Event()
+    lock_executing = Lock()
+    await lock_executing.acquire()
+
+    def block_execution(x, event, lock):
+        event.set()
+        with lock:
+            raise RuntimeError()
+
+    fut1 = c.submit(
+        block_execution,
+        1,
+        executing,
+        lock_executing,
+        resources={"A": 1},
+    )
+
+    await executing.wait()
+    fut2 = c.submit(inc, 1, resources={"A": 1})
+
+    while fut2.key not in a.tasks:
+        await asyncio.sleep(0.1)
+
+    fut1.release()
+    while a.tasks[fut1.key].state == "executing":
+        await asyncio.sleep(0.1)
```

Comment:
```suggestion
        await asyncio.sleep(0.01)
```
Shouldn't fut1 disappear from a entirely? Isn't there a race condition here?